pkg/alertmanager/multitenant.go

```diff
@@ -1089,6 +1186,39 @@ func (am *MultitenantAlertmanager) getPerUserDirectories() map[string]string {
 	return result
 }
 
+// UpdateState implements the Alertmanager service.
+func (am *MultitenantAlertmanager) ReadState(ctx context.Context, req *alertmanagerpb.ReadStateRequest) (*alertmanagerpb.ReadStateResponse, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id found in context")
+	}
+
+	am.alertmanagersMtx.Lock()
+	defer am.alertmanagersMtx.Unlock()
+	userAM, ok := am.alertmanagers[userID]
+	if !ok {
+		level.Debug(am.logger).Log("msg", "user does not have an alertmanager in this instance", "user", userID)
+		return &alertmanagerpb.ReadStateResponse{
+			Status: alertmanagerpb.READ_USER_NOT_FOUND,
+			Error:  "alertmanager for this user does not exists",
+		}, nil
+	}
+
+	state, err := userAM.getFullState()
+	if err != nil {
+		level.Error(am.logger).Log("msg", "failed to get full state", "user", userID, "err", err)
```

Comment:
This is already logged in the caller. Should we log it here as well?

Reply:
Good question. It's often useful to see failed requests on the server side too, but perhaps a `Debug` level message is good enough, to keep the verbosity down?